platform/core/src/services/MeasurementService/MeasurementService.js

```diff
@@ -0,0 +1,444 @@
+      version,
+    };
+    source.addOrUpdate = (definition, measurement) => {
+      return this.addOrUpdate(source, definition, measurement);
+    };
+    source.getAnnotation = (definition, measurementId) => {
+      return this.getAnnotation(source, definition, measurementId);
+    };
+
+    log.warn(`New '${name}@${version}' source added.`);
+    this.sources[id] = source;
+
+    return source;
+  }
+
+  /**
+   * Add a new measurement matching criteria along with mapping functions.
+   *
+   * @param {MeasurementSource} source Measurement source instance
+   * @param {string} definition Definition of the measurement (Annotation Type)
+   * @param {MatchingCriteria} matchingCriteria The matching criteria
+   * @param {Function} toSourceSchema Mapping function to source schema
+   * @param {Function} toMeasurementSchema Mapping function to measurement schema
+   * @return void
+   */
+  addMapping(
+    source,
+    definition,
+    matchingCriteria,
+    toSourceSchema,
+    toMeasurementSchema
+  ) {
+    if (!this._isValidSource(source)) {
+      log.warn('Invalid source. Exiting early.');
+      return;
+    }
+
+    if (!matchingCriteria) {
+      log.warn('Matching criteria not provided. Exiting early.');
+      return;
+    }
+
+    if (!toSourceSchema) {
+      log.warn('Source mapping function not provided. Exiting early.');
+      return;
+    }
+
+    if (!toMeasurementSchema) {
+      log.warn('Measurement mapping function not provided. Exiting early.');
+      return;
+    }
+
+    const mapping = {
+      matchingCriteria,
+      definition,
+      toSourceSchema,
+      toMeasurementSchema,
+    };
+
+    if (Array.isArray(this.mappings[source.id])) {
+      this.mappings[source.id].push(mapping);
+    } else {
+      this.mappings[source.id] = [mapping];
+    }
+
+    log.warn(`New measurement mapping added to source '${this._getSourceInfo(source)}'.`);
+  }
+
+  /**
+   * Get annotation for specific source.
+   *
+   * @param {MeasurementSource} source Measurement source instance
+   * @param {string} definition The source definition
+   * @param {string} measurementId The measurement service measurement id
+   * @return {Object} Source measurement schema
+   */
+  getAnnotation(source, definition, measurementId) {
+    if (!this._isValidSource(source)) {
+      log.warn('Invalid source. Exiting early.');
+      return;
+    }
+
+    if (!definition) {
+      log.warn('No source definition provided. Exiting early.');
+      return;
+    }
+
+    const measurement = this.getMeasurement(measurementId);
+    const sourceMappings = this.mappings[source.id];
+    const sourceMappingsByDefinition = sourceMappings.filter(
+      mapping => mapping.definition === definition
+    );
+
+    const matchedCriteriaMapping = sourceMappingsByDefinition.find(
+      ({ matchingCriteria }) => {
+        return (
+          measurement.points &&
+          measurement.points.length === matchingCriteria.points
+        );
+      },
+    );
```

Comment:
This is perfect for now. Curious to see how complex this gets as we contemplate adding things like... `isClosedShape`, angle between points, less than X points, more than X points, etc.

Reply:
I had started looking at something like this: https://www.npmjs.com/package/@hapi/joi
The requirements feel similar to unit test asserts

Reply:
Yeah, joi is really helpful asserting the scheme/data, i had used this before for mongo/node apps to validate params/body.